WalletWasabi/Tor/Socks5/TorSocks5ClientFactory.cs

```diff
@@ -0,0 +1,408 @@
+				|| ex.Message.StartsWith("Connection refused")) // Linux && OSX
+			{
+				error = ex;
+			}
+			// "No connection could be made because the target machine actively refused it" for non-English Windows.
+			catch (SocketException ex) when (ex.ErrorCode == 10061)
+			{
+				error = ex;
+			}
+			// "Connection refused" for non-English Linux.
+			catch (SocketException ex) when (ex.ErrorCode == 111)
+			{
+				error = ex;
+			}
+			// "Connection refused" for non-English OSX.
+			catch (SocketException ex) when (ex.ErrorCode == 61)
+			{
+				error = ex;
+			}
+			catch
+			{
+				// Ignored, since error is null.
+			}
+
+			return error is { };
+		}
+
+		/// <summary>
+		/// Sends a command to the Tor Socks5 connection and reads a response.
+		/// </summary>
+		/// <param name="sendBuffer">Sent data</param>
+		/// <param name="receiveBufferSize">Optionally, number of bytes expected to be received in the reply.</param>
+		/// <param name="cancellationToken">Cancellation token to cancel sending.</param>
+		/// <returns>Reply</returns>
+		/// <exception cref="TorResponseException">When we receive no response from Tor or the response is invalid.</exception>
+		private async Task<byte[]> SendAndReceiveAsync(TcpClient tcpClient, byte[] sendBuffer, int? receiveBufferSize = null, CancellationToken cancellationToken = default)
+		{
+			Guard.NotNullOrEmpty(nameof(sendBuffer), sendBuffer);
+
+			try
+			{
+				NetworkStream stream = tcpClient.GetStream();
+
+				// Write data to the stream.
+				await stream.WriteAsync(sendBuffer.AsMemory(0, sendBuffer.Length), cancellationToken).ConfigureAwait(false);
+				await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+
+				// If receiveBufferSize is null, zero or negative or bigger than TcpClient.ReceiveBufferSize
+				// then work with TcpClient.ReceiveBufferSize
+				var tcpReceiveBuffSize = tcpClient.ReceiveBufferSize;
+				var actualReceiveBufferSize = receiveBufferSize is null || receiveBufferSize <= 0 || receiveBufferSize > tcpReceiveBuffSize
+					? tcpReceiveBuffSize
+					: (int)receiveBufferSize;
+
+				// Receive the response.
+				var receiveBuffer = new byte[actualReceiveBufferSize];
+
+				// Read exactly "receiveBufferSize" bytes.
+				if (receiveBufferSize != null)
+				{
+					if (await stream.ReadExactlyAsync(receiveBuffer, receiveBufferSize.Value, cancellationToken).ConfigureAwait(false))
+					{
+						return receiveBuffer;
+					}
+
+					throw new TorResponseException($"Failed to read {receiveBufferSize.Value} bytes as expected from Tor SOCKS5.");
+				}
+
+				int receiveCount = await stream.ReadAsync(receiveBuffer.AsMemory(0, actualReceiveBufferSize), cancellationToken).ConfigureAwait(false);
+
+				if (receiveCount <= 0)
+				{
+					throw new TorResponseException($"Not connected to Tor SOCKS5 proxy: {TorSocks5EndPoint}.");
+				}
+
+				// If we could fit everything into our buffer, then return it.
+				if (!stream.DataAvailable)
```

Comment:
@lontivero Is this actually correct? This is unmodified code but my concern is that this IF statement can be true even when there are more data to read just because the counter-party is slow with sending. I have actually filed an issue to dotnet/runtime to ask for this specifically.

Reply:
CAn you link the issue that you created, just for the record?

Reply:
Sure: https://github.com/dotnet/runtime/issues/46122

Reply:
I saw the answer and it is what i would have expected, after all `DataAvailable` is just a way to ask the socket if there is something to read. This is a unusual way to read from a stream.

Reply:
So my analysis is as follows: The line 373 and the rest of the method reads ONLY reply from Tor as defined in RFC1928 (https://tools.ietf.org/html/rfc1928) of the following form:

![image](https://user-images.githubusercontent.com/58662979/102618545-49aecb00-413b-11eb-9f58-630edaa02fdf.png)

Currently we do that in one go which is not entirely OK because it may fail sometimes. I would say it fails very rarely if ever but it's still probably a bug. The correct solution for this would be to read the response byte by byte (unfortunately the length is variable) to get the response.

I'm not sure whether a fix for this should be included in this PR or in a follow-up PR. It should be safe to implement the fix because even if there are various Tor clients all should support SOCKS5 protocol, otherwise the implementations are broken. 

Any opinions on this?

Reply:
I agree. It is not for this PR.